torchdata/datapipes/iter/transform/callable.py

```diff
@@ -585,3 +591,200 @@ def __new__(
+        **threadpool_kwargs,
+    ) -> None:
+        super().__init__()
+        self.datapipe = source_datapipe
+
+        _check_unpickable_fn(fn)
+        self.fn = fn  # type: ignore[assignment]
+
+        if scheduled_tasks <= 0:
+            raise ValueError("'scheduled_tasks' is required to be a positive integer.")
+        self.scheduled_tasks = scheduled_tasks
+        if max_workers is not None and max_workers <= 0:
+            raise ValueError("'max_workers' is required to be a positive integer.")
+        self.max_workers = max_workers
+        self.threadpool_kwargs = threadpool_kwargs
+
+        self.input_col = input_col
+        if input_col is None and output_col is not None:
+            raise ValueError("`output_col` must be None when `input_col` is None.")
+        if isinstance(output_col, (list, tuple)):
+            if len(output_col) > 1:
+                raise ValueError("`output_col` must be a single-element list or tuple")
+            output_col = output_col[0]
+        self.output_col = output_col
+        validate_input_col(fn, input_col)
+
+    def _apply_fn(self, data):
+        if self.input_col is None and self.output_col is None:
+            return self.fn(data)
+
+        if self.input_col is None:
+            res = self.fn(data)
+        elif isinstance(self.input_col, (list, tuple)):
+            args = tuple(data[col] for col in self.input_col)
+            res = self.fn(*args)
+        else:
+            res = self.fn(data[self.input_col])
+
+        # Copy tuple to list and run in-place modification because tuple is immutable.
+        if isinstance(data, tuple):
+            t_flag = True
+            data = list(data)
+        else:
+            t_flag = False
+
+        if self.output_col is None:
+            if isinstance(self.input_col, (list, tuple)):
+                data[self.input_col[0]] = res
+                for idx in sorted(self.input_col[1:], reverse=True):
+                    del data[idx]
+            else:
+                data[self.input_col] = res
+        else:
+            if self.output_col == -1:
+                data.append(res)
+            else:
+                data[self.output_col] = res
+
+        # Convert list back to tuple
+        return tuple(data) if t_flag else data
+
+    def __iter__(self) -> Iterator[T_co]:
+        try:
+            executor = futures.ThreadPoolExecutor(max_workers=self.max_workers, **self.threadpool_kwargs)
+            futures_deque: deque = deque()
+            has_next = True
+            itr = iter(self.datapipe)
+            for _ in range(self.scheduled_tasks):
+                try:
+                    futures_deque.append(executor.submit(self._apply_fn, next(itr)))
+                except StopIteration:
+                    has_next = False
+                    break
+
+            # Yield must be hidden in closure so that the futures are submitted
+            # before the first iterator value is required.
```

Comment:
I don't fully understand this part. The futures should have been submitted to `executor`, right? Do you mean Python has not switched context to threadpool?

Reply:
TBH I don't fully understand it either why hiding yield is necessary (or is it? see below).

In `executor.map()` it totally makes sense to schedule tasks before calling `next` on the iterator.
When I don't hide it in a closure the first element is only yielded once `scheduled_tasks` many tasks have been completed.

At least I thought that's what happened as time until first increased (by a lot) with when increasing `scheduled_tasks` but I cannot reproduce it right now. Will probably change it back.

Reply:
This is potentially the problem of python threading. It needs a way to switch context as all threads share the same GIL. Normally we can do `time.sleep(0)` in the main thread to make the background thread picking up jobs.

Reply:
Changed it back to a generator as I could not reproduce the time issues. Currently without `time.sleep(0)`. Where should I add this or is it not really necessary?

Reply:
Using my simple function from before:
```python
def wait_a_bit(x):
    time.sleep(random.random() * 3)
    return x * 10
```
with random.seed(0) before every test
first `random.random() * 3` is `2.533265554575144`

we have without sleep:
| threads | scheduled_tasks | time until first   |
|---------|-----------------|--------------------|
| 64      | 10              | 2.540492057800293  |
| 64      | 64              | 2.537590265274048  |
| 64      | 100             | 2.5425333976745605 |
| 64      | 500             | 2.540599822998047  |
| 64      | 1000            | 2.5444140434265137 |


| threads | scheduled_tasks | time until 5000    |
|---------|-----------------|--------------------|
| 64      | 64              | 207.91916251182556 |
| 64      | 128             | 121.00023198127747 |
| 64      | 256             | 120.81454801559448 |
| 64      | 1024            | 120.81783604621887 |
| 64      | 2048            | 120.80706715583801 |
| 64      | 4096            | 120.81476616859436 |

with `time.sleep(0)`:

| threads | scheduled_tasks | time until first   |
|---------|-----------------|--------------------|
| 64      | 10              | 2.538611888885498  |
| 64      | 64              | 2.539947032928467  |
| 64      | 100             | 2.5512337684631348 |
| 64      | 500             | 2.5357728004455566 |
| 64      | 1000            | 2.537018299102783  |

| threads | scheduled_tasks | time until 5000    |
|---------|-----------------|--------------------|
| 64      | 64              | 207.94583225250244 |
| 64      | 128             | 121.03096175193787 |
| 64      | 256             | 120.7996666431427  |
| 64      | 1024            | 120.8266670703888  |
| 64      | 2048            | 120.84870266914368 |
| 64      | 4096            | 120.82779574394226 |

Essentially no difference. 

One thing to note though is that when the `executor` is being closed he will first finish all scheduled tasks. The only way to prevent this is to not use `with` and explicitly call `shutdown` and cancel tasks manually or use a newly added parameter `cancel_futures` (new in 3.9). tbh I do think that this is non-issue as usually one will probably exhaust the previous dp anyway and not stop in the middle?

Reply:
Thanks for confirming `sleep(0)` gives no impact on the perf. Then, let's keep it as it is.